File: cowichan/baSrcPaper/cilk/chain/par/outer.cpp
/*
 * outer: outer product
 *
 * input:
 *   winnow_points: a vector of (x, y) points
 *   nelts: the number of points
 *
 * output:
 *   outer_matrix: a real matrix, whose values are filled with inter-point
 *     distances
 *   outer_vector: a real vector, whose values are filled with origin-to-point
 *     distances
 */

#include <cilk/cilk.h>
#include <math.h>
#include <stdlib.h>
#include <stdio.h>
#include <string.h>

double *outer_matrix;
double *outer_vector;

typedef struct sPoint {
  int v, i, j;
} Point;

extern Point *winnow_points;

double sqr(double x) {
  return x * x;
}

double distance(Point a, Point b) {
  return sqrt(sqr(a.i - b.i) + sqr(a.j - b.j));
}

static double max(double a, double b) {
  return a > b ? a : b;
}

// parallel for on [begin, end)
static void fill_matrix(int begin, int end, int ncols) {
  int middle = begin + (end - begin) / 2;
  double nmax = -1;
  int j;

  if (begin + 1 == end) {
    for (j = 0; j < ncols; j++) {
      if (begin != j) {
        outer_matrix[begin*ncols + j] = distance(winnow_points[begin], winnow_points[j]);
        nmax = max(nmax, outer_matrix[begin*ncols + j]);
      }
      outer_matrix[begin*ncols + begin] = nmax * ncols;
      outer_vector[begin] = distance((Point){0, 0, 0}, winnow_points[begin]);
    }
    return;
  }
  cilk_spawn fill_matrix(begin, middle, ncols);
  cilk_spawn fill_matrix(middle, end, ncols);
}

void outer(int nelts) {
  outer_matrix = (double*) malloc (sizeof(double) * nelts * nelts);
  outer_vector = (double*) malloc (sizeof(double) * nelts);

  cilk_spawn fill_matrix(0, nelts, nelts);
}

File: cowichan/baSrcPaper/go/chain/par/src/all/outer.go
/*
 * outer: outer product
 *
 * input:
 *   vector: a vector of (x, y) points
 *   nelts: the number of points
 *
 * output:
 *   Outer_matrix: a real matrix, whose values are filled with inter-point
 *     distances
 *   Outer_vector: a real vector, whose values are filled with origin-to-point
 *     distances
 */

package all

import (
  "math"
)

type Double float64;

var Outer_matrix [][]Double;
var Outer_vector []Double;

func max(a, b Double) Double {
  if a > b {
    return a;
  }
  return b;
}

func sqr(x Double) Double {
  return x * x;
}

func distance(a, b Point) Double {
  return Double(math.Sqrt(float64(sqr(Double(a.i - b.i)) +
        sqr(Double(a.j - b.j)))));
}

func fill_matrix_impl(begin, end, ncols int, done chan bool) {
  if (begin + 1 == end) {
    var nmax Double = -1;
    for j := 0; j < ncols; j++ {
      if (begin != j) {
        Outer_matrix[begin][j] = distance(Winnow_points[begin], Winnow_points[j]);
        nmax = max(nmax, Outer_matrix[begin][j]);
      }
    }
    Outer_matrix[begin][begin] = Double(ncols) * nmax;
    Outer_vector[begin] = distance(Point{i: 0, j: 0}, Winnow_points[begin]);
    done <- true
  } else {
    middle := begin + (end - begin) / 2;
    go fill_matrix_impl(begin, middle, ncols, done);
    fill_matrix_impl(middle, end, ncols, done);
  }
}

func fill_matrix(nrows, ncols int) {
  done := make(chan bool);
  // parallel for on rows
  go fill_matrix_impl(0, nrows, ncols, done);
  for i := 0; i < nrows; i++ {
    <-done;
  }
}

func Outer(nelts int) {
  Outer_matrix = make ([][]Double, nelts)
  for i := range Outer_matrix {
    Outer_matrix [i] = make ([]Double, nelts)
  }

  Outer_vector = make ([]Double, nelts)

  fill_matrix(nelts, nelts);
}
